analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs

```diff
@@ -92,12 +94,52 @@ private static SyntaxNode OutOfOrderPlaceholderValue(MessageTemplatesParser.Plac
         return null;
     }
 
-    private static bool? MatchesName(string placeholderName, SyntaxNode placeholderValue) =>
+    private static bool? MatchesName(string placeholderName, SyntaxNode placeholderValue, bool isStrict) =>
         placeholderValue switch
         {
-            MemberAccessExpressionSyntax memberAccess => MatchesName(placeholderName, memberAccess.Name),
-            CastExpressionSyntax cast => MatchesName(placeholderName, cast.Expression),
-            NameSyntax name => new string(placeholderName.Where(char.IsLetterOrDigit).ToArray()).Equals(name.GetName(), StringComparison.OrdinalIgnoreCase),
+            MemberAccessExpressionSyntax memberAccess => MatchesName(placeholderName, memberAccess.Name, isStrict),
+            CastExpressionSyntax cast => MatchesName(placeholderName, cast.Expression, isStrict),
+            NameSyntax name => SimpleStringMatches(placeholderName, name.GetName(), isStrict),
             _ => null
         };
+
+    private static bool SimpleStringMatches(string placeholderName, string argumentName, bool isStrict)
+    {
+        if (isStrict)
+        {
+            return OnlyLetters(placeholderName).Equals(OnlyLetters(argumentName), StringComparison.OrdinalIgnoreCase);
+        }
+        else
+        {
+            var placeholderComponents = SplitByCamelCase(placeholderName);
+            var argumentComponents = SplitByCamelCase(argumentName);
+            return Array.Exists(placeholderComponents, x => Array.Exists(argumentComponents, y => x.Equals(y, StringComparison.OrdinalIgnoreCase)));
+        }
+    }
+
+    private static string[] SplitByCamelCase(string text)
+    {
+        var components = new List<string>();
+        var builder = new StringBuilder(text.Length);
+        foreach (var ch in text)
+        {
+            if (char.IsUpper(ch) && builder.Length > 0)
+            {
+                components.Add(builder.ToString());
+                builder.Clear();
+            }
+            if (char.IsLetter(ch))
+            {
+                builder.Append(ch);
+            }
+        }
+        if (builder.Length > 0)
+        {
+            components.Add(builder.ToString());
+        }
+        return [.. components];
```

Comment:
Isn't this copy the list contents to a new array? Why not return the list directly?